### PR TITLE
feat: add support for sum_discinct, count_distinct, approx_count_distinct

### PR DESCRIPTION
**Pull request by bcallender**
feat: add support for sum_discinct, count_distinct, approx_count_distinct

fix behavior to be more in line with pyspark semantics

### CODE REVIEW

**Comment by bcallender** [comment]
* **#249** <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/249?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#247** <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#243** <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/243?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#245** <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/245?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#242** <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/242?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/typedef-ai/fenic/242?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/fenic/api/functions/builtin.py [hunk 1]
```diff
@@ -53,6 +56,29 @@ def sum(column: ColumnOrName) -> Column:
...
+
+    Args:
+        column: Column or column name to compute the sum of distinct values
+
+    Returns:
+        A Column expression representing the sum distinct aggregation
+
+    Example:
+        ```python
+        df.group_by("k").agg(sum_distinct("v").alias("sum_unique"))
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py]: Nit: it would be nice to have an example input and output like they have in the spark documentation

File: src/fenic/api/functions/builtin.py [hunk 2]
```diff
@@ -147,6 +173,37 @@ def count(column: ColumnOrName) -> Column:
...
+    Args:
+        *cols: One or more columns or column names to include in the distinct count.
+
+    Returns:
+        A Column expression representing the count distinct aggregation over the provided columns.
+
+    Example:
+        ```python
+        # Single column
+        df.group_by("k").agg(count_distinct("v").alias("num_unique"))
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py]: same comment about showing example input and output, here and on approx_count_distinct

File: src/fenic/api/functions/builtin.py (hunk 1 above)
Comment (bcallender) [reply src/fenic/api/functions/builtin.py]: I was literally thinking the same thing and pushed a version with an example as you were commenting 😂

File: tests/_backends/local/dataframe/test_aggregations.py [hunk 3]
```diff
@@ -421,3 +426,86 @@ def test_stddev_aggregation(local_session):
...
+    assert result.schema["cd"] in (pl.Int64, pl.UInt32, pl.Int32)
+    assert result.schema["acd"] in (pl.Int64, pl.UInt32, pl.Int32)
+    assert result.schema["cd_text"] in (pl.Int64, pl.UInt32, pl.Int32)
+
+    a_row = result.filter(pl.col("group") == "A").row(0)
+    b_row = result.filter(pl.col("group") == "B").row(0)
+
+    # For group A: value cycles 0..9 (10 distinct, no nulls)
+    assert a_row[result.columns.index("cd")] == 10
+    assert a_row[result.columns.index("acd")] == 10
```
Comment (YoungVor) [inline tests/_backends/local/dataframe/test_aggregations.py]: curious about what size approx/HyperLogLog++ will start returning an approximation instead of an exact match.  I guess something between 2000 and 100,000 rows, based on the pyspark example

File: src/fenic/api/functions/builtin.py [hunk 4]
```diff
@@ -164,6 +291,57 @@ def collect_list(column: ColumnOrName) -> Column:
         ListExpr(Column._from_col_or_name(column)._logical_expr)
     )
 
+@validate_call(config=ConfigDict(strict=True, arbitrary_types_allowed=True))
+def approx_count_distinct(column: ColumnOrName) -> Column:
+    """Aggregate function: returns an approximate count (HyperLogLog++) of distinct non-null values.
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py:296]: kind of annoying polars doens't let you parameterize the std deviation.  Might be worth mentioning that as a limitation here since spark does allow it

File: src/fenic/api/functions/builtin.py [hunk 5]
```diff
@@ -164,6 +291,57 @@ def collect_list(column: ColumnOrName) -> Column:
         ListExpr(Column._from_col_or_name(column)._logical_expr)
     )
 
+@validate_call(config=ConfigDict(strict=True, arbitrary_types_allowed=True))
+def approx_count_distinct(column: ColumnOrName) -> Column:
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py:295]: Any reason this can't be a list of columns, and do the same null masking logic in expr_converter before doing the hyperlog part?  The pyspark documentation is a little unclear but i think it says it allows multiple inputs

File: tests/_backends/local/dataframe/test_aggregations.py (hunk 3 above)
Comment (YoungVor) [reply tests/_backends/local/dataframe/test_aggregations.py]: dont know if its worth a unit test with a large enough dataset

File: src/fenic/api/functions/builtin.py (hunk 5 above)
Comment (bcallender) [reply src/fenic/api/functions/builtin.py:295]: I believe `approx_count_distinct` only takes one column?

```
Parameters
colColumn or column name
The label of the column to count distinct values in.
```

Comment (bcallender) [reply src/fenic/api/functions/builtin.py:295]: Oh i see what you're saying:
```
This aggregate function returns a new Column, which estimates the approximate distinct count of elements in a specified column or a group of columns.
```

that is...really weird, since it definitely only takes one column as an input, it's not `*cols`, it's just `col`.

Comment (bcallender) [reply src/fenic/api/functions/builtin.py:295]: I think it's just a misnomer in the Pyspark 4 docs. The pyspark 3.x docs say: 
```
Aggregate function: returns a new Column for approximate distinct count of column col.
```

File: tests/_backends/local/dataframe/test_aggregations.py (hunk 3 above)
Comment (bcallender) [reply tests/_backends/local/dataframe/test_aggregations.py]: actually, even as low as 1000 rows I see it using the approximation. I added a unit test that shows the approx and exact counts with the relative error, it is typically very small but noticeable.

File: src/fenic/api/functions/builtin.py (hunk 4 above)
Comment (bcallender) [reply src/fenic/api/functions/builtin.py:296]: added a note to the docstring

File: src/fenic/api/functions/builtin.py (hunk 5 above)
Comment (YoungVor) [reply src/fenic/api/functions/builtin.py:295]: ah, fair enough, I just read the comment and didn't look closely at the signature

File: tests/_backends/local/dataframe/test_aggregations.py (hunk 3 above)
Comment (YoungVor) [reply tests/_backends/local/dataframe/test_aggregations.py]: sweet

File: tests/_backends/local/dataframe/test_aggregations.py [hunk 6]
```diff
@@ -421,3 +426,172 @@ def test_stddev_aggregation(local_session):
...
+
+        # Verify exact count is correct
+        assert exact == cardinality, f"{description}: exact count should be {cardinality}"
+
+        # Calculate relative error
+        relative_error = abs(approx - exact) / exact
+        print(f"{description}: relative error {relative_error:.2%}")
+
+        # HyperLogLog++ typical error rate is around 1.15% with default settings
+        # We allow up to 5% to be safe (accounting for edge cases)
```
Comment (YoungVor) [inline tests/_backends/local/dataframe/test_aggregations.py]: curious what edge cases would increase the error rate beyond the default HyperLogLog++ setting?

File: tests/_backends/local/dataframe/test_aggregations.py [hunk 7]
```diff
@@ -421,3 +426,172 @@ def test_stddev_aggregation(local_session):
...
+        _ = df.group_by("k").agg(sum_distinct(struct("v", "k"))).to_polars()
+
+
+def test_approx_count_distinct_approximation(local_session: Session):
+    """Test that approx_count_distinct actually uses approximation with HyperLogLog++.
+
+    This test verifies that the approximation is close to the exact count (within expected error bounds)
+    """
+    # Test with different cardinalities to see approximation behavior
+    test_cases = [
```
Comment (YoungVor) [inline tests/_backends/local/dataframe/test_aggregations.py]: tiny nit - you could make this a parameter to the test so each cardinatlity is a different unit test (up to you, not a big deal)

File: src/fenic/api/functions/builtin.py [hunk 8]
```diff
@@ -164,6 +291,58 @@ def collect_list(column: ColumnOrName) -> Column:
...
+    """Aggregate function: returns an approximate count (HyperLogLog++) of distinct non-null values.
+
+    Args:
+        column: Column or column name to approximately count distinct values in
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py]: should mention this only supports primitive scalars and arrays of primitives

File: src/fenic/core/_logical_plan/signatures/aggregate.py [hunk 9]
```diff
@@ -100,6 +103,36 @@ def register_aggregate_signatures():
         return_type=DoubleType
     ))
 
+    # Distinct aggregate functions
+    # approx_count_distinct - supports primitive scalars and arrays of primitives
+    FunctionRegistry.register("approx_count_distinct", FunctionSignature(
+        function_name="approx_count_distinct",
+        type_signature=OneOf([
+            Numeric(1),
+            ArrayOfPrimitives(1),
```
Comment (YoungVor) [inline src/fenic/core/_logical_plan/signatures/aggregate.py]: Should we allow embeddings?

File: src/fenic/core/_logical_plan/signatures/aggregate.py [hunk 10]
```diff
@@ -100,6 +103,36 @@ def register_aggregate_signatures():
         return_type=DoubleType
     ))
 
+    # Distinct aggregate functions
+    # approx_count_distinct - supports primitive scalars and arrays of primitives
```
Comment (YoungVor) [inline src/fenic/core/_logical_plan/signatures/aggregate.py]: can we not allow structs with only primitive fields?

it seems like the polars limitation is around nested arrays or nested structs

File: tests/_backends/local/dataframe/test_aggregations.py (hunk 7 above)
Comment (bcallender) [reply tests/_backends/local/dataframe/test_aggregations.py]: good call, changed.

File: tests/_backends/local/dataframe/test_aggregations.py (hunk 6 above)
Comment (bcallender) [reply tests/_backends/local/dataframe/test_aggregations.py]: i would think just massive cardinality? but i'm not sure.

File: src/fenic/api/functions/builtin.py (hunk 8 above)
Comment (bcallender) [reply src/fenic/api/functions/builtin.py]: updated to show it accepts all non-struct columns.

File: src/fenic/api/functions/builtin.py [hunk 11]
```diff
@@ -53,6 +56,57 @@ def sum(column: ColumnOrName) -> Column:
     )
 
 
+@validate_call(config=ConfigDict(strict=True, arbitrary_types_allowed=True))
+def sum_distinct(column: ColumnOrName) -> Column:
+    """Aggregate function: returns the sum of distinct numericvalues in the specified column.
```
Comment (YoungVor) [inline src/fenic/api/functions/builtin.py]: nit: space

File: tests/_backends/local/dataframe/test_aggregations.py [hunk 12]
```diff
@@ -421,3 +430,212 @@ def test_stddev_aggregation(local_session):
...
+    assert exact == 10_000
+
+    # Approximation should be within 5% of exact
+    relative_error = abs(approx - exact) / exact
+    assert relative_error <= 0.05, (
+        f"Approximation error {relative_error:.2%} exceeds 5% "
+        f"(exact={exact}, approx={approx})"
+    )
+
+def test_datatype_compatibility_with_count_distinct(local_session: Session):
```
Comment (YoungVor) [inline tests/_backends/local/dataframe/test_aggregations.py:603]: awesome, thank you!

**Review by YoungVor — APPROVED** [review]
looks great!
